File: articles/linear-algebra-for-systems-modeling/articles/sparse-matrices-and-computational-efficiency/calculators/r/sparse_matrix_efficiency_calculator.R
result <- data.frame(
  calculator = "sparse_matrices_and_computational_efficiency_calculator",
  model_name = "synthetic_sparse_matrix_efficiency_audit",
  matrix_dimension = 250,
  nonzero_entries = 1244,
  density = 0.019904,
  dense_storage_mb = 0.5,
  coordinate_storage_mb_estimate = 0.019904,
  storage_reduction_factor = 25.12,
  average_row_degree = 3.98,
  max_row_degree = 6,
  isolated_rows = 0,
  matrix_vector_product_norm = 31.6,
  iterative_residual_initial = 15.8,
  iterative_residual_final = 0.06,
  iterations = 60,
  warning = "Sparse matrix metrics depend on zero interpretation, density, storage format, thresholding, solver behavior, conditioning, and validation context."
)

dir.create("outputs", recursive = TRUE, showWarnings = FALSE)
write.csv(result, "outputs/r_sparse_matrices_and_computational_efficiency_calculator.csv", row.names = FALSE)
print(result)

File: articles/linear-algebra-for-systems-modeling/articles/sparse-matrices-and-computational-efficiency/calculators/python/sparse_matrix_efficiency_calculator.py
from __future__ import annotations

import csv
import json
from pathlib import Path


def main() -> None:
    output_dir = Path("outputs")
    output_dir.mkdir(parents=True, exist_ok=True)

    result = {
        "calculator": "sparse_matrices_and_computational_efficiency_calculator",
        "model_name": "synthetic_sparse_matrix_efficiency_audit",
        "matrix_dimension": 250,
        "nonzero_entries": 1244,
        "density": 0.019904,
        "dense_storage_mb": 0.5,
        "coordinate_storage_mb_estimate": 0.019904,
        "storage_reduction_factor": 25.12,
        "average_row_degree": 3.98,
        "max_row_degree": 6,
        "isolated_rows": 0,
        "matrix_vector_product_norm": 31.6,
        "iterative_residual_initial": 15.8,
        "iterative_residual_final": 0.06,
        "iterations": 60,
        "warning": "Sparse matrix metrics depend on zero interpretation, density, storage format, thresholding, solver behavior, conditioning, and validation context."
    }

    with (output_dir / "sparse_matrices_and_computational_efficiency_calculator.json").open("w", encoding="utf-8") as handle:
        json.dump(result, handle, indent=2, sort_keys=True)

    with (output_dir / "sparse_matrices_and_computational_efficiency_calculator.csv").open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=list(result.keys()))
        writer.writeheader()
        writer.writerow(result)

    print(json.dumps(result, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()
